src/routers/alpha-router/alpha-router.ts

```diff
@@ -1645,6 +1652,105 @@ export class AlphaRouter
+    const shouldEnableCachedRoutesCacheInvalidationFix =
+      Math.random() * 100 <
+      (this.cachedRoutesCacheInvalidationFixRolloutPercentage ?? 0);
+
+    // we have to write cached routes right before checking swapRouteRaw is null or not
+    // because getCachedRoutes in routing-api do not use the blocks-to-live to filter out the expired routes at all
+    // there's a possibility the cachedRoutes is always populated, but swapRouteFromCache is always null, because we don't update cachedRoutes in this case at all,
+    // as long as it's within 24 hours sliding window TTL
+    if (shouldEnableCachedRoutesCacheInvalidationFix) {
+      // theoretically, when routingConfig.intent === INTENT.CACHING, optimisticCachedRoutes should be false
+      // so that we can always pass in cachedRoutes?.notExpired(await blockNumber, !routingConfig.optimisticCachedRoutes)
+      // but just to be safe, we just hardcode true when checking the cached routes expiry for write update
+      // we decide to not check cached routes expiry in the read path anyway
+      if (!cachedRoutes?.notExpired(await blockNumber, true)) {
+        // optimisticCachedRoutes === false means at routing-api level, we only want to set cached routes during intent=caching, not intent=quote
+        // this means during the online quote endpoint path, we should not reset cached routes
+        if (routingConfig.intent === INTENT.CACHING) {
+          // due to fire and forget nature, we already take note that we should set new cached routes during the new path
+          newSetCachedRoutesPath = true;
+          metric.putMetric(`SetCachedRoute_NewPath`, 1, MetricLoggerUnit.Count);
+
+          // from above, there's a chance that swapRouteFromChain is populated, if we start to filter cached routes by blocks-to-live
+          // then in that case we already know that swapRouteFromChainPromise will resolve to swapRouteFromChain
+          // otherwise, we know that swapRouteFromChain didn't get populated, so we need to fetch it here.
+          // high level idea is that, in alpha-router class, this.getSwapRouteFromChain should only get awaited/resolved once, due to its expensiveness.
+          // this method relies on getCachedRoutes method can filter by blocks-to-live
+          const swapRouteFromChainAgainPromise = swapRouteFromChain
+            ? swapRouteFromChainPromise
+            : this.getSwapRouteFromChain(
+                amount,
+                currencyIn,
+                currencyOut,
+                protocols,
+                quoteCurrency,
+                tradeType,
+                routingConfig,
+                v3GasModel,
+                v4GasModel,
+                mixedRouteGasModel,
+                gasPriceWei,
+                v2GasModel,
+                swapConfig,
+                providerConfig
+              );
+
+          // intentionally a fire and forget, because getSwapRouteFromChain can be expensive
```

Comment:
this only sets new one if it needs to update, but actually does not delete old bad ones right?

Reply:
i mean its an in-place update in dynamo, because the partition key doesnt change, its still https://github.com/Uniswap/routing-api/blob/08646b11ca1246ba4c666f4909661e234520a1ee/lib/handlers/router-entities/route-caching/model/pair-trade-type-chain-id.ts#L29

`${this.currencyIn}/${this.currencyOut}/${this.tradeType}/${this.chainId}`

Reply:
but you mentioned there are multiple routes for the same key right? THerefore, if we dont go inside and invalidate some of these routes then it wont get fixed?

Reply:
i see what you saying..

right now the cached routes data structure is that all the routes for the same key will have the same block number that they were cached, [here](https://github.com/Uniswap/smart-order-router/blob/main/src/providers/caching/route/model/cached-routes.ts#L28). but i can see that we should keep each block number, for each route that we stored into this cached routes object, and only invalidate some routes from this cached routes object. 

is that what you are asking? if so, i def agree in principal, but i can elaborate on some implementation difficulties due to how cached routes are implemented in routing-api..

Reply:
yeah... i think thats the right way to go about it, otherwise we lose all routes which is also bad if only one of them is failing right?

Reply:
we discussed offline. current cached routes implementation is sub-optimal, but with time constraint, as well as next gen in mind, we decide to just keep the key level (tokenin/tokenout/tradetype/chainid) to update the cached routes in dynamo. ideally we know we want to update each individual cached routes with its latest block number updated.